authentik/stages/email/stage.py

```diff
@@ -172,10 +178,52 @@ def get_challenge(self) -> Challenge:
     def challenge_valid(self, response: ChallengeResponse) -> HttpResponse:
         return super().challenge_invalid(response)
 
+    def _get_cache_key(self) -> str:
+        """Return the cache key used for rate limiting email recovery attempts."""
+        user = self.get_pending_user()
+        user_email_hashed = sha256(user.email.lower().encode("utf-8")).hexdigest()
+        return EMAIL_RECOVERY_CACHE_KEY + user_email_hashed
+
+    def _is_rate_limited(self) -> int | None:
+        """Check whether the email recovery attempt should be rate limited.
+
+        If the request should be rate limited, update the cache and return the
+        remaining time in minutes before the user is allowed to try again.
+        Otherwise, return None."""
+        cache_key = self._get_cache_key()
+        attempts = cache.get(cache_key, [])
+
+        stage = self.executor.current_stage
+        stage.refresh_from_db()
+        max_attempts = stage.recovery_max_attempts
+        cache_timeout = stage.recovery_cache_timeout
+
+        now = time.time()
+        start_window = now - cache_timeout
+        recent_attempts_in_window = [attempt for attempt in attempts if attempt > start_window]
+
+        if len(recent_attempts_in_window) >= max_attempts:
+            retry_after = int(min(recent_attempts_in_window) + cache_timeout - now)
+            minutes_left = max(1, math.ceil(retry_after / 60))
+            return minutes_left
+
+        recent_attempts_in_window.append(now)
+        cache.set(cache_key, recent_attempts_in_window, cache_timeout)
+
+        return None
+
     def challenge_invalid(self, response: ChallengeResponse) -> HttpResponse:
+        if minutes_left := self._is_rate_limited():
+            error = _(
+                "Too many account recovery attempts. Please try again after {minutes} minutes."
```

Comment:
```suggestion
                "Too many account verification attempts. Please try again after {minutes} minutes."
```

Email stage doesn't have to be used for recovery, can be used for validation and other things too